src/org/labkey/test/components/ui/grids/QueryGrid.java

```diff
@@ -299,13 +355,181 @@ public QueryGrid clearAllSelections()
+            String buttonText = "Undo\nSave";
+            headerText = headerText.contains(buttonText) ? headerText.substring(0, headerText.indexOf(buttonText)) : headerText;
+            String appButtonText = "UndoSave";
+            viewName = headerText.contains(appButtonText) ? headerText.substring(0, headerText.indexOf(appButtonText)) : headerText;
+        }
+        else
+        {
+            viewName = "";
+        }
+
+        return viewName.trim();
+    }
+
+    /**
+     * Click the Undo button in the header. Will wait for the grid to update.
+     *
+     * @return This grid.
+     */
+    public QueryGrid clickUndoButton()
+    {
+        WebElement undoButton = Locator.buttonContainingText("Undo").findElement(elementCache().panelHeader());
+
+        // Wait for the grid to update.
+        doAndWaitForUpdate(undoButton::click);
+
+        return this;
+    }
+
+    /**
+     * Is the 'Undo' button visible on the grid.
+     *
+     * @return True if visible, false otherwise.
+     */
+    public boolean isUndoButtonVisible()
+    {
+        return Locator.buttonContainingText("Undo").findWhenNeeded(elementCache().panelHeader()).isDisplayed();
+    }
+
+    /**
+     * Click the Save (View) button in the grid header. If the grid is showing the default view clicking save will show
+     * a {@link SaveGridViewDialog}, otherwise clicking save will save the changes to the view currently applied to the grid.
+     *
+     * @param expectSaveDialog Indicate if a 'Save View' dialog is expected. Should only happen if you are saving the default view.
+     * @return A {@link SaveGridViewDialog} or null if a saved dialog is not expected.
+     */
+    public SaveGridViewDialog clickSaveButton(boolean expectSaveDialog)
+    {
+        WebElement saveButton = Locator.buttonContainingText("Save").findElement(elementCache().panelHeader());
+        saveButton.click();
+
+        if(expectSaveDialog)
+        {
+            return new SaveGridViewDialog(getDriver(), this);
+        }
+        else
+        {
+            return null;
+        }
+    }
+
+    /**
+     * Click the 'Save as...' drop down button menu option to change the view name. The 'Save as...' option will not be
+     * visible if the default view is being changed.
+     *
+     * @return A {@link SaveGridViewDialog}
+     */
+    public SaveGridViewDialog clickSaveAsButton()
+    {
+        BootstrapMenu bootstrapMenu = new BootstrapMenu(getDriver(), elementCache().panelHeader());
+        bootstrapMenu.clickSubMenu(false, "Save as...");
+
+        return new SaveGridViewDialog(getDriver(), this);
+    }
+
+    /**
+     * Is the 'Save' button visible on the grid.
+     *
+     * @return True if visible, false otherwise.
+     */
+    public boolean isSaveButtonVisible()
+    {
+        return Locator.buttonContainingText("Save").findWhenNeeded(elementCache().panelHeader()).isDisplayed();
```

Comment:
`Locator.isDisplayed`
```suggestion
        return Locator.buttonContainingText("Save").isDisplayed(elementCache().panelHeader());
```

Reply:
Changed.